designs/pluggable-message-routing/README.md

```diff
@@ -0,0 +1,263 @@
+}
+```
+
+##### Scenario: custom cloning behaviour
+
+The initial logic around what messages get cloned does not change - the node
+identifies if cloning needs to occur based on how many nodes it is wired to and
+how many messages it is asked to send in a single go.
+
+Rather than do the cloning, the node will set the `cloneMessage` property to indicated
+whether cloning should occur or not.
+
+The actual cloning would happen between the `preRoute` and `onSend` steps.
+
+This would allow a `preRoute` handler to do its own cloning behaviour and then set
+`cloneMessage` to `false` so that no further cloning would happen for that message.
+
+One very important aspect to highlight is that messages will not be cloned until
+after the `preRoute` stage - so any modifications of the message will potentially
+mutate multiple messages. If earlier handlers want to modify the message in any
+way, they will have to honour the `cloneMessage` flag and clone the message themselves.
+
+
+##### Scenario: remote message routing
+
+The `preRoute` step can be used to do remote message routing. As no message cloning
+has happened by this point, it avoids that overhead when serialising the message
+to send over the network serves the same purpose.
+
+All steps, up to and including `onReceive` will have a way to tell the runtime to
+stop processing that message and to not pass it on to any later steps. This would
+allow a `preRoute` handler to decide a message must be sent to a remote runtime
+and that no further local processes was needed.
+
+##### Scenario: flow debugger - breakpoints
+
+A breakpoint could be triggered when a node sends a message or receives one.
+
+To break on node send, it would either register a handler on the `preSend` step
+(if it was necessary to break before *any* of the messages sent in that one call
+are processed) or `preRoute` if it works on a single message. That's a question
+for the Flow Debugger design and doesn't need addressing now. However, the options
+are there.
+
+To break on node receive, it would register on `preReceive`.
+
+In either case, if the breakpoint is triggered, the flow debugger needs to be able
+to indefinitely halt the passing of messages.
+
+This informs us that all of the handlers need to be able to complete asynchronously -
+so the flow debugger can defer sending on any messages until it is resumed. It would
+also allow the flow debugger to do step-debugging, by registering handlers at every
+step.
+
+##### Scenario: flow testing
+
+The flow testing design introduces the ability to:
+ - verify the contents/structure of a message being passed to a node
+ - verify the contents/structure of a message send by a node
+ - stub out an entire node with custom test-case-specific behaviour
+
+That maps to the steps:
+ - `onReceive` - verify message passed to a node
+ - `preSend` - verify the message(s) sent by a node
+ - `preRoute` - pass the message to a Test Stub node instead of the real node
+
```

Comment:
while this is for testing - could it also answer the itch re runtime type checking  if suitable schemas could be provided "everywhere" to  check against ?

Reply:
Anyone could write a hook that does anything. So yes, a dedicated message validation hook could well be created.